### PR TITLE
Add AllBitsSet DIM

### LINKED ISSUE #69676: [API Proposal]: Add AllBitsSet to the IBinaryNumber Interface
### Background and motivation

Conceptually, `T.AllBitsSet` returns an instance of the binary number type `T` where every bit is set to 1. For example, `Int32.AllBitsSet` is `0xFFFFFFFF`, or `-1`. This concept is generally useful for masking purposes. For example, Vectors use `AllBitsSet` often for Conditional Select, as bitwise-anding a number with `AllBitsSet` will produce the same number, in contrast to bitwise-anding a number with `Zero`. As brought up by https://github.com/dotnet/runtime/issues/60882, `AllBitsSet` would be a nice addition to the `IBinaryNumber` interface, potentially enabling future generic refactoring of `Scalar<T>`.

The following types would explicitly implement `AllBitsSet`: `Double`, `Half`, `NFloat`, `Single`, `Byte`, `Char`, `Int16`, `Int32`, `Int64`, `IntPtr`, `SByte`, `UInt16`, `UInt32`, `UInt64`, `UIntPtr`, `BigInteger`, `Int128`, `Uint128`.

Design of the implementation for all of these types is trivial except for `BigInteger`. For `BigInteger`, the most obvious implementation would be to return `MinusOne`, as despite the internal bit representation not being literally `AllBitsSet` (due to `BigInteger` having unbounded size), it behaves as such for bitwise operations.

### API Proposal

```csharp
namespace System.Numerics
{

  public partial interface IBinaryNumber<TSelf>
  {
       static abstract TSelf AllBitsSet { get; }
  }

}
```

### API Usage

```csharp

T genericAlgorithm(T input) 
where T : IBinaryNumber<T> {
    return input + T.AllBitsSet;
}
```

### Alternative Designs

When deciding whether a method is implemented explicitly or implicitly for Generic Math, we have used the following rule: if the concept is trivially handled by knowing the concrete type, it's implemented explicitly. In other words, it is not exposed on the concrete type. This way, the implementation is only available when working with the generic interface. 

In this case, we follow this logic and implement `AllBitsSet` explicitly, as knowing the concrete type leads to a trivial implementation (https://source.dot.net/#System.Private.CoreLib/Scalar.cs,11). However, it could be argued that `AllBitsSet` is conceptually tricky enough for some of these types that a implicit implementation is warranted.

### Risks

Currently, refactoring `Scalar<T>` to take advantage of this is not possible without a language change. More specifically, the language doesn't have a way to "bridge" type constraints. So, since `Vector<T>` is `where T : struct` but we'd want `Scalar<T>` to be `where T : struct, IBinaryNumber<T>`, we'd need the language to provide some way to allow `Vector<T>` to use `Scalar<T>` despite the mismatching type requirements.

### PR DESCRIPTION
**Pull request by dakersnar**
Fixes https://github.com/dotnet/runtime/issues/69676

Continuation of effort here: https://github.com/dotnet/runtime/pull/71356

I wanted to get this PR up early so it can be merged as soon as https://github.com/dotnet/linker/issues/2865 is fixed.

Is there an easy way to unit test DIMs? I can add unit tests if that would be helpful/possible.

### CODE REVIEW

**Comment by dakersnar** [comment]
Seems like the bug was actually fixed, so this should be ready to merge

File: src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs [hunk 1]
```diff
@@ -11,7 +11,7 @@ public interface IBinaryNumber<TSelf>
         where TSelf : IBinaryNumber<TSelf>
     {
         /// <summary>Gets an instance of the binary type in which all bits are set.</summary>
-        static abstract TSelf AllBitsSet { get; } // TODO: add the DIM once https://github.com/dotnet/linker/issues/2865 is fixed
+        static virtual TSelf AllBitsSet => ~TSelf.Zero;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs:14]: Can we add tests for this?  Hopefully we have tests for all other DIMs, but if not, we should add some for those, too.

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 2]
```diff
@@ -10485,7 +10485,7 @@ public partial interface IBinaryInteger<TSelf> : System.IComparable, System.ICom
...
     {
-        static abstract TSelf AllBitsSet { get; }
+        static virtual TSelf AllBitsSet => throw null;
```
Comment (stephentoub) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:10488]: Nit: I don't think this is how the genapi tool would generate this, is it?  I'd expect it to be more like:
```suggestion
        static virtual TSelf AllBitsSet { throw null; }
```

File: src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs (hunk 1 above)
Comment (dakersnar) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs:14]: Sure, I'll look into this.

File: src/libraries/System.Runtime/tests/System.Runtime.Tests.csproj [hunk 3]
```diff
@@ -111,6 +112,7 @@
     <Compile Include="System\NullableMetadataTests.cs" />
     <Compile Include="System\NullableTests.cs" />
     <Compile Include="System\NullReferenceExceptionTests.cs" />
+    <Compile Include="System\Numerics\IBinaryNumberTests.cs" />
```
Comment (tannergooding) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests.csproj:115]: nit: The other generic math files are under a special item group down on L262

File: src/libraries/System.Runtime/tests/System/Numerics/GenericMathDimHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        BinaryNumberDimHelper(int value)
+        {
+            this.value = value;
+        }
+        static BinaryNumberDimHelper INumberBase<BinaryNumberDimHelper>.Zero => new BinaryNumberDimHelper(0);
+
+        static BinaryNumberDimHelper IBitwiseOperators<BinaryNumberDimHelper, BinaryNumberDimHelper, BinaryNumberDimHelper>.operator ~(BinaryNumberDimHelper value)
+        {
+            return new BinaryNumberDimHelper(~value.value);
+        }
```
Comment (tannergooding) [inline src/libraries/System.Runtime/tests/System/Numerics/GenericMathDimHelpers.cs:24]: nit: Fix newlines and add `private` before the constructor:
```suggestion
    public struct BinaryNumberDimHelper : IBinaryNumber<BinaryNumberDimHelper>
    {
        public int Value = 0;

        private BinaryNumberDimHelper(int value)
        {
            Value = value;
        }

        static BinaryNumberDimHelper INumberBase<BinaryNumberDimHelper>.Zero => new BinaryNumberDimHelper(0);

        static BinaryNumberDimHelper IBitwiseOperators<BinaryNumberDimHelper, BinaryNumberDimHelper, BinaryNumberDimHelper>.operator ~(BinaryNumberDimHelper value)
        {
            return new BinaryNumberDimHelper(~value.value);
        }
```

File: src/libraries/System.Runtime/tests/System/Numerics/IBinaryNumberTests.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+using Xunit;
+
+namespace System.Numerics.Tests
+{
+
+    public class IBinaryNumberTests
+    {
+
+        [Fact]
+        public static void AllBitsSetTest()
```
Comment (tannergooding) [inline src/libraries/System.Runtime/tests/System/Numerics/IBinaryNumberTests.cs:18]: nit: Fix newlines and remove unnecessary usings
```suggestion
using Xunit;

namespace System.Numerics.Tests
{
    public sealed class IBinaryNumberTests
    {
        [Fact]
        public static void AllBitsSetTest()
```

**Review by tannergooding — APPROVED** [review]

File: src/libraries/System.Runtime/tests/System.Runtime.Tests.csproj [hunk 6]
```diff
@@ -261,6 +261,8 @@
 
   <ItemGroup>
     <Compile Include="$(CommonTestPath)System\GenericMathHelpers.cs" Link="Common\System\GenericMathHelpers.cs" />
+    <Compile Include="System\Numerics\GenericMathDimHelpers.cs" />
+    <Compile Include="System\Numerics\IBinaryNumberTests.cs" />
     <Compile Include="System\ByteTests.GenericMath.cs" />
     <Compile Include="System\CharTests.GenericMath.cs" />
     <Compile Include="System\DecimalTests.GenericMath.cs" />
```
Comment (dakersnar) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests.csproj:268]: @tannergooding does this placement look correct now?

Comment (tannergooding) [reply src/libraries/System.Runtime/tests/System.Runtime.Tests.csproj:268]: LGTM.

**Pull request merged**